demos/object_detection_demo_ssd_async/detection_model.h

```diff
@@ -0,0 +1,51 @@
+/*
+// Copyright (C) 2018-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+#pragma once
+#include "model_base.h"
+#include "opencv2/core.hpp"
+
+class DetectionModel :
+    public ModelBase
+{
+public:
+    /// Constructor
+    /// @param modelFileName name of model to load
+    /// @param confidenceThreshold - threshold to eleminate low-confidence detections.
+    /// Any detected object with confidence lower than this threshold will be ignored.
+    /// @param useAutoResize - if true, image will be resized by IE.
+    /// Otherwise, image will be preprocessed and resized using OpenCV routines.
+    /// @param labels - array of labels for every class. If this array is empty or contains less elements
+    /// than actual classes number, default "Label #N" will be shown for missing items.
+    DetectionModel(const std::string& modelFileName, float confidenceThreshold, bool useAutoResize, const std::vector<std::string>& labels);
+
+    virtual void preprocess(const InputData& inputData, InferenceEngine::InferRequest::Ptr& request, MetaData*& metaData) override;
```

Comment:
```suggestion
    virtual void preprocess(const InputData& inputData, InferenceEngine::InferRequest::Ptr& request, std::unique_ptr<MetaData>& metaData) override;
```

The caller is supposed to take ownership of the metadata, so it should be a `unique_ptr`.

Reply:
changed to shared_ptr as Metadata is packed inside shared_ptr later (to capture safely inside lambdas)

Reply:
You should still use `unique_ptr` for the parameter, as it captures the semantics better. Using `shared_ptr` implies that `preprocess` can retain a reference to the metadata, and I don't think that's intended. You can easily move the `unique_ptr` into a `shared_ptr` in the caller.

Reply:
that's too many conversions and overprotection. I think that's extra effort

Reply:
no relevant any more, please close